### PR TITLE
Design doc: training and validation

### LINKED ISSUE #390: When the model training is completed, a method is needed to test the training results.
The solution I want:
Add a testing action or function in training SQL
```sql
SELECT *
FROM iris.train
TRAIN DNNClassifier

...
TESTING iris.test
...

LABEL class
INTO sqlflow_models.my_dnn_model;

```

Log addition:
```
Training set accuracy: 0.96721
Testing  set accuracy: 0.90132
Done training
```

### PR DESCRIPTION
**Pull request by weiguoz**
fix #390

### CODE REVIEW

File: doc/training_and_validation.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+Validation is used to evaluate the trained model, like accuracy rating, overfitting, etc. In SQLFlow, automation is the core value. So SQLFlow proposed to do training and validation in one SQL. It requires splitting the user-specific data into two parts: training data and validation data correspond to two processes(train and validation) respectively. 
+
+Notice, we talk about the **train** process in this post.
+
+## Overall
+
+SQLFlow generates a temporary table following the user-specific dataset, trains and evaluates a model.
+
+<img src="/Users/weiguo/go/src/github.com/sql-machine-learning/sqlflow/doc/figures/training_and_validation.png" width="60%">
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: typo: broken link.

File: doc/training_and_validation.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| <data> | <data> | <data> |
+| ...    |        |        |
+
+We want to split the result into 80% training data and 20% validation data.
+
+**We add a column sqlflow_random via RAND() and save the result to a temporary table.** Note the `RAND()` function returns a random number between 0 (inclusive) and 1. The result temporary table looks like the following.
+
+| temp_table |        |        |                |
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: Markdown doesn't support merged cells, HTML does:
<table>
  <tr>
    <th colspan="4">temp_table</th>
  </tr>
  <tr>
    <td>col1</td>
    <td>col2</td>
    <td>col3</td>
    <td>train_val_split</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.3</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.9</td>
  </tr>
  <tr>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>&lt;data&gt;</td>
    <td>0.5</td>
  </tr>
  <tr>
    <td>...</td>
    <td></td>
    <td></td>
    <td></td>
  </tr>
</table>

File: doc/training_and_validation.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
+# Design: Training and Validation
+
+Validation is used to evaluate the trained model, like accuracy rating, overfitting, etc. In SQLFlow, automation is the core value. So SQLFlow proposed to do training and validation in one SQL. It requires splitting the user-specific data into two parts: training data and validation data correspond to two processes(train and validation) respectively.
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: Maybe the intro could be a little bit more concise like:

"A common ML training job usually involves two kinds of datasets: training data and validation data. These two datasets will be generated automatically by SQLFlow through randomly splitting the select results."

File: doc/training_and_validation.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+```go
+func runExtendedSQL(slct string, db *DB, pr *extendedSelect, modelDir string) *PipeReader { 
+  // ...
+  if os.Getenv("SQLFLOW_submitter") == "alps" {
+    pr = split
+    return submitALPS(wr, pr, db, cwd) // training & validation in a same function
+  }
+  
+  if pr.train {
+    train(pr, slct, db, cwd, wr, modelDir)           // using pr.training
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: Training and validation belong to one Python program. The program's pseudo code looks like

```python
for step in range(10000):
    model.train(batch_data)
    if step % 1000 == 0: # validation will be run 10 times in a training job
        model.eval(val_dataset)
```

File: doc/training_and_validation.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+| temp_table |        |        |                |
+| ---------- | ------ | ------ | -------------- |
+| col1       | col2   | col3   | sqlflow_random |
+| <data>     | <data> | <data> | 0.3            |
+| <data>     | <data> | <data> | 0.9            |
+| <data>     | <data> | <data> | 0.5            |
+| ...        |        |        |                |
+
+We can generate the corresponding SQL using the following code template
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md:71]: 1. Maybe elaborate on how we generate the temporary table name. Something like `fmt.Sprintf("temp_table%d", rand.Int())`.
1. Which database does the temporary table belong to?

File: doc/training_and_validation.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+   ```go
+   type extendedSelect struct {
+     // ...
+     training   standardSelect // training dataset
+     validation standardSelect // validation dataset
+   }
+   ```
+
+2. Split `runExtendedSQL` to `train & validate` on training
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: No need to split runExtendedSQL. Training and validation belong to one Python program.

File: doc/training_and_validation.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  // ...
+}
+```
+
+`validate` returns the evaluation of the trained model.
+
+## Codegen
+
+For tensorflow submitter, we generate `evaluation` code to evaluate the model. 
+
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: Add a drop table section.

File: doc/training_and_validation.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  }
+  // ...
+}
+```
+
+`validate` returns the evaluation of the trained model.
+
+## Codegen
+
+For tensorflow submitter, we generate `evaluation` code to evaluate the model.
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: For TensorFlow submitter, we generate training dataset and validation dataset according to `extendedSelect.training` and `extendedSelect.validation`.

File: doc/training_and_validation.md (hunk 7 above)
Comment (weiguoz) [reply doc/training_and_validation.md]: Done.

File: doc/training_and_validation.md (hunk 5 above)
Comment (weiguoz) [reply doc/training_and_validation.md:71]: 1. I agree with generating an elaborate name for the temporary table. Because multi-users run SQLFlow with their own isolated dataset at the same time.

Comment (weiguoz) [reply doc/training_and_validation.md:71]: 2.1 For common databases, like MySQL. SQLFlow creates a temporary table with the attribute [TEMPORARY ](https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html), and we don't need to specify an existed database name.
I also made a test in hive: `CREATE TEMPORARY TABLE what_ever_tbl AS SELECT *, rand() AS rdm FROM iris.train;`, It works.

2.2 For Maxcompute.  SQLFlow saves the temporary table into the `current project`  which must be specified by the user.

Plus, we do not have to test if the temporary table exists for common databases. Let me add some details to this doc.

File: doc/training_and_validation.md [hunk 9]
```diff
@@ -0,0 +1,109 @@
+# Design: Training and Validation
+
+A common ML training job usually involves two kinds of datasets: training data and validation data. These two datasets will be generated automatically by SQLFlow through randomly splitting the select results.
+
+## Overall
+SQLFlow generates a temporary table following the user-specific dataset, trains and evaluates a model.
```
Comment (Yancey0623) [inline doc/training_and_validation.md]: database?

File: doc/training_and_validation.md [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+    SELECT *, RAND() AS sqlflow_random FROM (
+        {.StandardSQL}
+    )
+```
+
+## How to split
+
+**We fetch the training/validation data using two different queries respectively.** 
+   
+The query for training data can be written as `SELECT * FROM temp_table WHERE sqlflow_random < 0.8`, which fetches row1 and row3 etc.. The query for validation data can be written as `SELECT * FROM temp_table WHERE sqlflow_random >= 0.8`, which fetches the rest of the rows.
```
Comment (Yancey0623) [inline doc/training_and_validation.md:96]: Maybe we can generate a different random number for each epoch, that can make the train/validation more exhaustive:
1. start an epoch and random range: [a, b];
1. train with `sqlflow_random < a and sqlflow_random > b`
1. validation with `sqlflow_random >= a and sqllfow_random <= b`
1. re-random [a1, b1] and back to step `1`

more methods about the cross-validation: https://en.wikipedia.org/wiki/Cross-validation_(statistics)

File: doc/training_and_validation.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+```SQL
+CREATE TABLE {.TempTableName} AS
+    SELECT *, RAND() AS sqlflow_random FROM (
+        {.StandardSQL}
+    )
+```
+
+### Naming temporary table
+- For common databases, like MySQL, Hive   
+SQLFlow creates the temporary table with the attribute [TEMPORARY ](https://dev.mysql.com/doc/refman/8.0/en/create-temporary-table.html), so we do not have to test if the temporary table exists.
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: I am afraid `TEMPORARY` doesn't work in our case. From the documentation

> You can use the `TEMPORARY` keyword when creating a table. A `TEMPORARY` table is visible only within the current session, and is dropped automatically when the session is closed. 

We create tables in Go and read the contents in Python. They are different sessions. Therefore the Python session can't see the temporary table created by Go.

File: doc/training_and_validation.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+- For Maxcompute  
+SQLFlow saves the temporary table into the current project which must be specified by the user. 
+Meanwhile, we specify the [LIFECYCLE](http://help.aliyun-inc.com/internaldoc/detail/27808.html?spm=a2c1f.8259794.2.1.64a896d5dR0z6r) to 1 day to release the temporary table automatically.
```
Comment (tonyyang-svail) [inline doc/training_and_validation.md]: Please change the link to the public doc: https://www.alibabacloud.com/help/doc-detail/55297.htm

Some training jobs take more than a day. How about setting it to 14 days? I vaguely remember the default lifecycle of a Maxcompute table is 14 days?

File: doc/training_and_validation.md (hunk 10 above)
Comment (tonyyang-svail) [reply doc/training_and_validation.md]: > Maybe we can generate a different random number for each epoch.

I am not sure it is common to change training&validation dataset during a training process, it lets the model see the validation data... The k-fold cross validation I learnt in college usually runs **the whole training process** k times.

File: doc/training_and_validation.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+## Release the temporary table
+In the end, we remove the temporary table by
+```SQL
+drop table if exists {temporary_table_name}
+```
+We created the temporary table specified with `TEMPORARY` or `LIFECYCLE`, but SQLFlow would rather release the temporary table explicitly.
+
+## Notes
+
+- If the column sqlflow_random already exists, SQLFlow chooses to quit
```
Comment (typhoonzero) [inline doc/training_and_validation.md:126]: Why not use this column directly, for cases that the dataset is already partitioned to training set and validation set like IMDB.

File: doc/training_and_validation.md (hunk 12 above)
Comment (weiguoz) [reply doc/training_and_validation.md]: `If no lifecycle is specified, the table or partition cannot be automatically recycled by MaxCompute.`
14 days is enough. Also, `lifecycle` can be adjusted.

File: doc/training_and_validation.md (hunk 11 above)
Comment (weiguoz) [reply doc/training_and_validation.md]: For common databases(excludes Maxcompute), I turned to save the temporary table to sqlflow_workspace(a database maintained by SQLFlow)

File: doc/training_and_validation.md (hunk 13 above)
Comment (weiguoz) [reply doc/training_and_validation.md:126]: I'm not sure if the existing sqlflow_random's data type or randomness will help to split.

File: doc/training_and_validation.md (hunk 10 above)
Comment (joyyoj) [reply doc/training_and_validation.md:96]: The method that Yancey1986 proposed here seems better. No tmp table needed.

File: doc/training_and_validation.md (hunk 13 above)
Comment (tonyyang-svail) [reply doc/training_and_validation.md:126]: I think @typhoonzero means the users should be able to specify which row to be used as training or validation by deliberately adding the sqlflow_random column.

So for SQLFlow, if sqlflow_random already exists, it should use it directly.

File: doc/training_and_validation.md (hunk 10 above)
Comment (tonyyang-svail) [reply doc/training_and_validation.md:96]: > we can generate a different random number for each epoch

@Yancey1989 I am not sure it is common to change training&validation dataset during a training process, it lets the model see the validation data... The k-fold cross validation I learned in college usually runs the whole training process k times.

> The method that Yancey1986 proposed here seems better. No tmp table needed.

@joyyoj  Looks like we still need to construct a temporary table by adding the sqlflow_random column.

**Review by tonyyang-svail — DISMISSED** [review]

**Review by tonyyang-svail — APPROVED** [review]
